### PR TITLE
[Dart] change to mini libraries

### LINKED ISSUE #3828: [REQ] [Dart] Avoid using part and create mini libraries instead
### Is your feature request related to a problem? Please describe.

I had to refactor the templates from using `part` to creating mini libraries because having models as parts caused problems. That was for my case but more generally, I think it might be a better and is the recommended approach.

## Describe the solution you'd like

Refactor the templates to avoid using `part` 

## Describe alternatives you've considered

I think there are arguments to be made both ways.  The official line is to [avoid using part and create mini libraries instead](https://dart.dev/guides/libraries/create-library-packages#organizing-a-library-package)

Using `part` does makes life easier in creating/editing generators as all `part` files are like one big file, but that does lose the potential benefits of having 'library private' access between files and can mean less control over what is ultimately exposed.

Also, using `part` for models and apis also can limit having an extra step with dart codegen that uses `part` (although that may be less of a concern for most people)

## Additional context

https://dart.dev/guides/libraries/create-library-packages
https://stackoverflow.com/a/27764138/1992736

### PR DESCRIPTION
**Pull request by nickmeinhold**
fixes #3828

Also a stepping stone on the way to having client and serializer options in a single generator. 

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [x] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.
@ircecho (2017/07) @swipesight (2018/09) @jaumard (2018/09)

### CODE REVIEW

**Comment by wing328** [comment]
Looks like some Dart client samples have not been updated:
```
Perform git status
On branch pull/4549
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   samples/client/petstore/dart-jaguar/flutter_petstore/openapi/lib/model/pet.dart
	modified:   samples/client/petstore/dart-jaguar/flutter_proto_petstore/openapi/lib/model/pet.proto
	modified:   samples/client/petstore/dart-jaguar/openapi/lib/model/pet.dart
	modified:   samples/client/petstore/dart-jaguar/openapi_proto/lib/model/pet.proto
	modified:   samples/client/petstore/dart2/openapi/lib/api.dart
	modified:   samples/client/petstore/dart2/openapi/lib/api/pet_api.dart
	modified:   samples/client/petstore/dart2/openapi/lib/api/store_api.dart
	modified:   samples/client/petstore/dart2/openapi/lib/api/user_api.dart
	modified:   samples/client/petstore/dart2/openapi/lib/api_client.dart
	modified:   samples/client/petstore/dart2/openapi/lib/api_exception.dart
	modified:   samples/client/petstore/dart2/openapi/lib/api_helper.dart
	modified:   samples/client/petstore/dart2/openapi/lib/auth/api_key_auth.dart
	modified:   samples/client/petstore/dart2/openapi/lib/auth/authentication.dart
	modified:   samples/client/petstore/dart2/openapi/lib/auth/http_basic_auth.dart
	modified:   samples/client/petstore/dart2/openapi/lib/auth/oauth.dart
	modified:   samples/client/petstore/dart2/openapi/lib/model/api_response.dart
	modified:   samples/client/petstore/dart2/openapi/lib/model/category.dart
	modified:   samples/client/petstore/dart2/openapi/lib/model/order.dart
	modified:   samples/client/petstore/dart2/openapi/lib/model/pet.dart
	modified:   samples/client/petstore/dart2/openapi/lib/model/tag.dart
	modified:   samples/client/petstore/dart2/openapi/lib/model/user.dart

no changes added to commit (use "git add" and/or "git commit -a")
Please run 'bin/utils/ensure-up-to-date' locally and commit changes (UNCOMMITTED CHANGES ERROR)
Exited with code 1
```

File: modules/openapi-generator/src/main/resources/dart2/api_helper.mustache [hunk 1]
```diff
@@ -54,7 +57,7 @@ String parameterToString(dynamic value) {
 /// Returns the decoded body by utf-8 if application/json with the given headers.
 /// Else, returns the decoded body by default algorithm of dart:http.
 /// Because avoid to text garbling when header only contains "application/json" without "; charset=utf-8".
-String _decodeBodyBytes(Response response) {
+String decodeBodyBytes(Response response) {
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/dart2/api_helper.mustache]: I notice that you removed the leading underscore so these methods are now officially public, right?

File: modules/openapi-generator/src/main/resources/dart2/model.mustache [hunk 2]
```diff
@@ -1,12 +1,8 @@
...
-{{>class}}
-{{/isEnum}}
-{{/model}}
-{{/models}}
+    {{#model}}
+        {{#imports}}
+            import '{{.}}';
+        {{/imports}}
+        {{#isEnum}}{{>enum}}{{/isEnum}}{{^isEnum}}{{>class}}{{/isEnum}}
+    {{/model}}
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/dart2/model.mustache]: Should we remove these leading spaces (line 2 to 7)?

File: modules/openapi-generator/src/main/resources/dart2/api_helper.mustache (hunk 1 above)
Comment (nickmeinhold) [reply modules/openapi-generator/src/main/resources/dart2/api_helper.mustache]: Yeah they've gone from being "library private" to "library public".  Ironically, this has made them less 'public' - because the package was using the "part of" directive, the whole package was effectively a single library unit meaning the methods were accessible anywhere.  Now the api_helper.dart has to be imported before they can be used.  Externally there is no change as the api_helper.dart file is not exported.

File: modules/openapi-generator/src/main/resources/dart2/model.mustache (hunk 2 above)
Comment (nickmeinhold) [reply modules/openapi-generator/src/main/resources/dart2/model.mustache]: Yeah good call, I'll make the change now.  My troubled local branch seems to have magically come good btw.

File: modules/openapi-generator/src/main/resources/dart2/api_helper.mustache (hunk 1 above)
Comment (wing328) [reply modules/openapi-generator/src/main/resources/dart2/api_helper.mustache]: Thanks for the explanation.

**Comment by nickmeinhold** [comment]
The failure on bitrise is unrelated to the PR right?  

For the circle failure, I ran `./bin/utils/ensure-up-to-date --batch` and I get some modified files in `samples/.../dart-jaguar` and some in `samples/.../dart2/openapi`.  There's also a bunch of groovy and kotlin sample stuff generated.  I don't understand what's going on here, some quick questions: 
- why does running ensure-up-to-date produce all these unrelated files? 
- where is the `samples/.../dart2/openapi` being generated?  Running `bin/dar2-petstore.sh` generates `samples/.../dart2/petstore_client_lib` so I don't understand where `samples/.../dart2/openapi` is coming from...? 

My local branch has now mysteriously returned to the state of being both ahead and behind the remote branch... I have no idea what's happening sorry.  @wing328 is it annoying for you to have to perform the update again?  I'm happy to start a new branch if that's easier.

File: samples/client/petstore/dart-jaguar/flutter_petstore/openapi/lib/model/pet.dart [hunk 3]
```diff
@@ -1,9 +1,9 @@
 import 'package:jaguar_serializer/jaguar_serializer.dart';
 
 
-import 'package:openapi/model/tag.dart';
+import 'package:openapi/model/package:openapi/model/tag/dart.dart';
```
Comment (wing328) [inline samples/client/petstore/dart-jaguar/flutter_petstore/openapi/lib/model/pet.dart]: @nickmeinhold as discussed, we'll probably need to update the dart jaguar template to fix this.

Please take your time and let me know if you need help.

Comment (nickmeinhold) [reply samples/client/petstore/dart-jaguar/flutter_petstore/openapi/lib/model/pet.dart]: Thanks @wing328, I've put up a fix but having trouble with the CI - I have already run `./bin/dart-jaguar-petstore.sh` and `./bin/utils/ensure-up-to-date --batch` locally, but neither produces any changes. It seems weird but this branch has been doing some really weird stuff.  I could start a new branch again, I don't see why that would help but maybe worth a try?

Comment (wing328) [reply samples/client/petstore/dart-jaguar/flutter_petstore/openapi/lib/model/pet.dart]: Before running `./bin/dart-jaguar-petstore.sh`, did you run `mvn clean package` to recompile the JAR? I think you'd but asking just to make sure.

Comment (wing328) [reply samples/client/petstore/dart-jaguar/flutter_petstore/openapi/lib/model/pet.dart]: Btw `./bin/utils/ensure-up-to-date --batch` (batch mode) is not ready yet and we've switched back to just `./bin/utils/ensure-up-to-date`

Comment (nickmeinhold) [reply samples/client/petstore/dart-jaguar/flutter_petstore/openapi/lib/model/pet.dart]: Ugh, no that was totally it. I guess multi-tasking does make you stupid ;-) Thanks @wing328, good catch.

I think maybe the failure on Travis is unrelated...? 

Thanks for the heads up re: batch mode.  I was using batch mode because it prompts you to when you run just `./bin/utils/ensure-up-to-date` - maybe that's been changed already?  I'll grab the latest master and check.

Comment (nickmeinhold) [reply samples/client/petstore/dart-jaguar/flutter_petstore/openapi/lib/model/pet.dart]: Just checked, running `./bin/utils/ensure-up-to-date` still says the following: 
```
IMPORTANT: this script should be run by the CI (e.g. Shippable) to ensure that the 'samples/' folder is up to date.
When running locally execute with argument --batch.
```
Want me to create an issue to change it?

File: samples/client/petstore/dart-jaguar/openapi/lib/model/pet.dart [hunk 4]
```diff
@@ -1,9 +1,9 @@
 import 'package:jaguar_serializer/jaguar_serializer.dart';
 
 
-import 'package:openapi/model/tag.dart';
+import 'package:openapi/model/tag/dart.dart';
```
Comment (wing328) [inline samples/client/petstore/dart-jaguar/openapi/lib/model/pet.dart]: Looks like this will be a problem as it should be just `tag.dart` instead.

File: samples/client/petstore/groovy/bin/main/org/openapitools/api/ApiUtils.groovy [hunk 5]
```diff
@@ -0,0 +1,54 @@
+package org.openapitools.api
```
Comment (wing328) [inline samples/client/petstore/groovy/bin/main/org/openapitools/api/ApiUtils.groovy]: I noticed some changes (newly-added files) for Groovy Petstore client. Which script did you run to update the Groovy Petstore client?

I run `./bin/groovy-petstore.sh` with the latest master but don't see any changes.

Please undo these changes if not needed.

**Comment by wing328** [comment]
No update so we're closing this one for the time being. Happy to reopen when it's ready for another review.
